functorch/csrc/BatchRulesReduceOps.cpp

```diff
@@ -275,6 +275,101 @@ std::tuple<Tensor, optional<int64_t>, Tensor, optional<int64_t>> aminmax_batchin
   return std::make_tuple(min, 0, max, 0);
 }
 
+std::tuple<Tensor, Tensor> cdist_batch_rule_helper(
+    const Tensor& x1, optional<int64_t> x1_bdim,
+    const Tensor& x2, optional<int64_t> x2_bdim) {
+
+  auto x1_ = moveBatchDimToFront(x1, x1_bdim);
+  auto x2_ = moveBatchDimToFront(x2, x2_bdim);
+
+  int64_t d1 = static_cast<int64_t>(x1_bdim.has_value());
+  int64_t d2 = static_cast<int64_t>(x2_bdim.has_value());
+  while (x2_bdim && x1_.dim() - d1 > x2_.dim() - d2) {
+    // x1:                      [  S1, S2, .............., SK]
+    // x2: [B, D1, ..., DN]  -> [B, 1,  1, ..., 1, D1 ..., DN]
+    x2_ = x2_.unsqueeze(1);
+  }
+  while (x1_bdim && x2_.dim() - d2 > x1_.dim() - d1) {
+    // x2:                      [  S1, S2, .............., SK]
+    // x1: [B, D1, ..., DN]  -> [B, 1,  1, ..., 1, D1 ..., DN]
+    x1_ = x1_.unsqueeze(1);
+  }
+  // Backward requires contiguous tensors
+  if (!x1_.is_contiguous()) {
+    x1_ = x1_.contiguous();
+  }
+  if (!x2_.is_contiguous()) {
+    x2_ = x2_.contiguous();
+  }
+  return std::make_tuple(x1_, x2_);
+}
+
+std::tuple<Tensor,optional<int64_t>> cdist_batch_rule(
+    const Tensor& x1, optional<int64_t> x1_bdim,
+    const Tensor& x2, optional<int64_t> x2_bdim,
+    const double p, c10::optional<int64_t> compute_mode) {
+
+  auto x12 = cdist_batch_rule_helper(x1, x1_bdim, x2, x2_bdim);
+  return std::make_tuple(at::cdist(std::get<0>(x12), std::get<1>(x12), p, compute_mode), 0);
+}
+
+std::tuple<Tensor,optional<int64_t>> cdist_forward_batch_rule(
+    const Tensor& x1, optional<int64_t> x1_bdim,
+    const Tensor& x2, optional<int64_t> x2_bdim,
+    const double p, c10::optional<int64_t> compute_mode) {
+
```

Comment:
We should probably validate that x1 and x2 have logical dimension >= 2. Otherwise we might throw a weird error in the batching rule

Reply:
By default, right now the error looks like that (if passing 2d tensor before batching):
```
RuntimeError: cdist only supports at least 2D tensors, X1 got: 1D
```
I think this is readable. What do you think ?

Code:
```python
def func(x1, x2):
    return torch.cdist(x1, x2, 1.0, 'use_mm_for_euclid_dist')

t1 = torch.rand(4, 5)
t2 = torch.rand(3, 5)

out = vmap(func, in_dims=(0, None))(t1, t2)
```

Reply:
Thank you for checking. Yes, that is readable -- let's leave the code as-is